compiler/rustc_resolve/src/imports.rs

```diff
@@ -1325,7 +1327,11 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
 
         // Skip if we are inside a named module (in contrast to an anonymous
         // module defined by a block).
-        if let ModuleKind::Def(..) = import.parent_scope.module.kind {
+        // Skip if the import is public or was used through non scope-based resolution,
+        // e.g. through a module-relative path.
+        if self.effective_visibilities.is_exported(self.local_def_id(id))
+            || import.used.get() == Some(Used::Other)
```

Comment:
```suggestion
            import.used.get() == Some(Used::Other) || self.effective_visibilities.is_exported(self.local_def_id(id))
```
Cheaper check first.